tests/utils/test_cache_utils.py

```diff
@@ -164,6 +165,56 @@ def _random_kvs(config):
         self.assertTrue(cached_keys.shape == (1, 1, 10, 128))
         self.assertTrue(cached_values.shape == (1, 1, 10, 128))
 
+    def test_static_cache_exportability(self):
+        """
+        Tests that static cache works with `torch.export()`
+        """
+        device = "cpu"
+        dtype = torch.float32
+        max_batch_size = 1
+
+        config = AutoConfig.from_pretrained(
+            "google/gemma-2b",
+            torch_dtype=dtype,
+            use_cache=True,
+        )
+        m = AutoModelForCausalLM.from_pretrained(
+            "google/gemma-2b",
+            config=config,
+            torch_dtype=dtype,
+            attn_implementation="sdpa",  # Export and ExecuTorch only works for SdpaAttention
+        ).to(device)
+        tokenizer = AutoTokenizer.from_pretrained("google/gemma-2b")
+        inputs = tokenizer(["The best color is"], return_tensors="pt").to(device)["input_ids"]
+
+        class ExportatibleModelWithStaticCache(torch.nn.Module):
```

Comment:
@ArthurZucker BTW, we don't need to address it in this PR, so it shouldn't block merging this PR. It's just to kick off another static cache related discussion for ExecuTorch since this code snippet is a good example it explain the context.

Reply:
@gante Let me explain more about the `generate` (inference) part, and give the full picture of how I imagine the collaboration/integration.

**Disclaimer**: Note that the ultimate goal of export and lower to [ExecuTorch](https://pytorch.org/executorch/main/index.html) is to run inference on edge devices just like onnx, tflite, etc. via `Optimum`.

## Why is the adapter forward() needed?
In [ExecuTorch](https://pytorch.org/executorch/main/index.html) we have a c++ runtime for LLMs that could load the exported transform model (a binary format `.pte`) for inference. To utilize that runtime, the `forward()` must comply with the same signature, which looks like:
```
    def forward(
        token: torch.Tensor,
        input_pos: Optional[torch.Tensor],
    ) -> torch.Tensor:
```

So basically in [the prototype PR](https://github.com/huggingface/transformers/pull/31706) I created this adapter `forward()` for Gemma-2b to make it compatible with that c++ runtime. With the adapter, after the model is exported, it can be loaded for inference by running cli command like this:
```
cmake-out/examples/models/llama2/llama_main --model_path=gemma.pte --tokenizer_path=gemma_tokenizer.bin --prompt="Hello world!"
```
Please note that the primary goal of creating the adapter forward() is to demonstrate the end-to-end workflow of exporting and lowering a Hugging Face model to ExecuTorch with minimal changes by reusing the c++ runtime. It doesn't mean we need to adapt to all Hugging Face models to it.

## How to generate in a more scalable way?
Of course ,it's not scale to add such an adapter `forward()` for all models. For users to be able to inference/generate using the export model, ideally the experience should be similar to eager or torch compiled model. To make it happen, ExecuTorch can provide a dedicated runtime that:
  1. comply with `forward()` of Hugging Face transformers `PreTrainedModel` *(With one exception. Explain it in next section)*
  2. can be used in eager python directly or via `Optimum` *(either directly implement in python like [this](https://github.com/pytorch/executorch/blob/main/examples/models/llama2/runner/generation.py) or expose the c++ implementation via pybind)*
With that dedicated runtime, such an adapter won't be needed.

## What is the requirement?
It's important to emphasize that to make the above approach work, there is a technical requirement must be complied. The requirement is: **The cache must be statically configured at the time of export.**

Today in HF transformers the cache is not configurable via `AutoConfig` *(static config used to construct a transformer model)*. To make it work with `torch.export()`, I have to statically instantiate the `StaticCache` in the adapter `forward()` as a workaround. That is also the other reason of having this adapter forward() in both [the prototype PR](https://github.com/huggingface/transformers/pull/31706) and in this unit test. As discussed with @amyeroberts yesterday, I'm proposing to add an option to make it statically configurable at model construct time.



I feel maybe it's better to make it a co-design proposal somewhere so we can iterate on it and loop in Optimum team. @amyeroberts @gante @ArthurZucker what would be the recommended place to repost it?

Reply:
(Writing here in case I forget, in advance of potentially moving the discussion somewhere else)

> Today in HF transformers the cache is not configurable via AutoConfig

@guangy10 This is actually something we want to change! At the moment, the `model` object holds:
1. `config`, which specifies the model architecture
2. `generation_config`, which specifies generation parameterization

We were thinking of creating a `cache_config` field within `generation_config`, which would fully parameterize a cache. I'm assuming this would solve the question, correct? If so, we can (and should) fast-track it 🤗

Reply:
The cache is actually already configurable via the AutoConfig API, because you can set parameters there that will be passed to the `generation_config`. As @gante mentioned having the `cache_config` alwayhs passed to the `generation_config` would solve the configurable at construct time constraint! 

We can open a PR for this if it not already the case!

Reply:
@guangy10 Feel free to open an issue in transformers and link it here (or visa versa). The optimum team will be able to comment and discuss there. cc @michaelbenayoun for the optimum side

Reply:
Opened in https://github.com/huggingface/transformers/issues/32253. Let's continue the discussion there.